### DISCUSSION TITLE
Questions about coupled variable and global parameter

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#18218] [Q&A General] [answered]
I made an application that has 3 variables and 3 kernels.

I would like to introduce a new variable and there is no kernel for this variable.

What I want to know is

**"Can coupled variable (introduced new variable) be changed by off-diagonal jacobian?"** 

Also, in this application, I set the global parameter (e.g. A=1000).

![image](https://user-images.githubusercontent.com/46009457/124028025-864c6080-d9c1-11eb-9217-cdbe31a290f6.png)

Like above, there is a pipe. Everywhere in pipe has constant A (A=1000), except for triangle (A=0).

So far, **A was constant number, however I want to change A at a certain condition(place). Is it possible through MOOSE?**

**Comment by ngrilli** [comment, marked as answer]
Dear @hokonkim 

The coupling of a variable inside a user defined kernel is described in example number 3:
https://github.com/idaholab/moose/tree/next/examples/ex03_coupling

If you specify better your equations and the terms you want to introduce, it is easier to help you.

If you want a different A parameter in different part of the geometry the best strategy is to create multiple "blocks",
this can be done in gmsh by specifying more "physical volumes".
"A" can be defined as a material property in a material object, then restrict the material to a certain block,
you will need two materials with A = 0 and A = 1000 in your simulation.
The material property can be imported into the kernel with getMaterialProperty, you can see many examples of this in this repository:
https://github.com/ngrilli/c_pfor_am

Best Regards,
Nicolò

**Reply by hokonkim** [reply]
@ngrilli Thanks for your reply.

If the geometry (pipe) consists of one volume (block), then what is the best way to implement this?

**Reply by GiudGiud** [reply]
You will want to use a function to define the spatial dependency **if** you want to only use one block

**Reply by hokonkim** [reply]
I tried to check "Functions" in Syntax, however still I have no idea about this.

Could you point me out for specific syntax?  or

Is there any uploaded input examples at MOOSE framework website?

**Reply by GiudGiud** [reply]
Just use a ParsedFunction for the defintion. Then a (AD?) GenericFunctionMaterial to take that parsed function
https://mooseframework.inl.gov/source/functions/MooseParsedFunction.html
https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html

**Reply by ngrilli** [reply]
Dear @hokonkim 

If you have no specific reason to use a single block, just add another block.
You need to create two "physical volumes" in gmsh.

There are plenty of examples online, for instance this one:
https://stackoverflow.com/questions/53914425/create-single-mesh-for-multiple-bodies-in-gmsh

Nicolò

**Reply by GiudGiud** [reply]
I fully agree with Nicolo that this is the more natural way to do this. 
Doing a single block is likely not to capture the true shape of that triangle, unless you are very careful with meshing

**Reply by hokonkim** [reply]
@ngrilli @GiudGiud 

I think it needs more detail explanation.

What I am trying to implement is degradation ("poisoning) happens in the fuel cell.

When degradation happens, somewhere in TPB (in previous discussion, I said the pipe) will be deactivated.

If somewhere in TPB is deactivated, then that point(element?) in TPB should be shut off or at least there is no more reaction contributing to a performance.

Also, in TPB, there is an only S0 constants related to TPB reaction. 

In terms of volumetric mesh, TPB is just one mesh.

So, I tried to find a way to give a zero S0 value to deactivated TPB  when TPB are poisoned.

**Reply by GiudGiud** [reply]
so you're trying to remove elements during the simulation essentially?
or just disable the reaction in those elements?

For the latter:
If you go the function route, you can make that function depend on time.
If you need more fine grained control you will have to create your own reaction kernel, and have the criteria for activation/deactivation of the reaction there.

**Reply by hokonkim** [reply]
@GiudGiud 

I think disabling the reaction in the elements is enough.

I've made a simple function in an input file.

But, **"how to disable the reaction in the element?". Are there any examples I can refer?**

I've made several reaction kernels as well, but still I need to think more about this if I want to use kernel...

**Can I control a reaction with probability?**
ex) when species A and B exist in (or near) an element, A + B -> C reaction occurs with 0.01% probability.

**PS. Can GlobalParams in input file override the value given at the source kernel?**
ex) (input file) GlobalParams/z = 8
     (source kernel) params.addParam<Real>("z", 4.0, "Electron number (num of electrons transferred)");
     In this case, z = 8 or z = 4?

**Reply by GiudGiud** [reply]
Ok good, that is very doable. What does the disabling depend on?
If it's just a profile that you are imposing and that you already know the time dependence of, then just use a function in space and time. Then use a reaction kernel with a function for the coefficent. 

If not that use a (material property + reaction kernel with coefficient material property) OR (reaction kernel with the coefficient hard-coded in the kernel)  and put in the criteria for disabling here. 

Theoretically speaking, you absolutely could use a probabilistic approach. I do not think it is required here, a continuous approach will be just as easy to follow, and will not necessitate an uncertainty study. 

GlobalParams can fill the parameters yes.
For example for this parameter you gave, the default of 4 is not used, the global parameter is used.
Now if you specify for that given block a value of 6 in the input block, then BOTH the default and the global parameters are overridden.

**Reply by GiudGiud** [reply]
An example of using a function is kernel is for this kernel
https://mooseframework.inl.gov/source/kernels/UserForcingFunction.html
